src/Provisioner.js

```diff
@@ -0,0 +1,341 @@
+const log = require('npmlog');
+const util = require('./util.js');
+
+const RemoteRoom = require("matrix-appservice-bridge").RemoteRoom;
+const MatrixRoom = require("matrix-appservice-bridge").MatrixRoom;
+
+const LEAVE_UNPROVIS_AFTER_MS = 10*60*1000;
+const ROOM_JOIN_TIMEOUT_MS = 1*60*1000;
+const DEFAULT_POWER_REQ = 50;
+
+class Provisioner {
+  constructor (bridge, twitter, config) {
+    this._bridge = bridge;
+    this._as = bridge.appService;
+    this._app = bridge.appService.app;
+    this._twitter = twitter;
+    this._config = config.provisioning;
+    if(this._config.required_power_level === undefined) {
+      this._config.required_power_level = DEFAULT_POWER_REQ;
+    }
+  }
+
+  init () {
+    if (this._config.enable && !(this._app.use && this._app.get && this._app.post)) {
+      log.error('Could not start provisioning.');
+      return;
+    }
+
+    if(!this._config.enable) {
+      log.info("Provisioner", "Disabled provisoning");
+      this._app.use((req, res, next) => {
+        // Disable all provision endpoints by not calling 'next' and returning an error instead
+        if (this.isProvisionRequest(req)) {
+          res.header("Access-Control-Allow-Origin", "*");
+          res.header("Access-Control-Allow-Headers",
+                "Origin, X-Requested-With, Content-Type, Accept");
+          res.status(500);
+          res.json({error: 'Provisioning is not enabled.'});
+        }
+        else {
+          next();
+        }
+      });
+      return;
+    }
+
+    log.info("Provisioner", "Enabled provisoning");
+
+    this._app.use((req, res, next) => {
+      //Forward the request onwards!
+      if (this.isProvisionRequest(req)) {
+        res.header("Access-Control-Allow-Origin", "*");
+        res.header("Access-Control-Allow-Headers",
+            "Origin, X-Requested-With, Content-Type, Accept");
+      }
+      next();
+    });
+
+    const manageLink = (req, res) => {
+      Promise.coroutine(this._requestWrap)(this, this._manageLink, req, res);
+    };
+
+    this._app.put("/_matrix/provision/:roomId/:type/:name", manageLink);
+    this._app.delete("/_matrix/provision/:roomId/:type/:name", manageLink);
+    this._app.get("/_matrix/provision/:roomId/links", (req, res) => {
+      Promise.coroutine(this._requestWrap)(this, this._listLinks, req, res);
+    });
+    this._app.get("/_matrix/provision/show/:screenName", (req, res) => {
+      Promise.coroutine(this._requestWrap)(this, this._queryProfile, req, res);
+    });
+  }
+
+  * _requestWrap (self, func, req, res) {
+    try {
+      const result = yield Promise.coroutine(func)(self, req);
+      if(result !== undefined) {
+        if(result.err) {
+          res = res.status(result.err);
+        }
+        res.json(result);
+      }
+    }
+    catch (err) {
+      res.status(500).json({error: "An internal error occured."});
+      log.error("Provisioner", "Error occured: %s", err);
```

Comment:
It seems `npmlog` doesn't quite handle this as you'd expect, printing the error out twice. I think the solution would be to remove the '%s', but I'm not sure about the double printing thing.
```JS
> let log = require('npmlog');
> let err = new Error('I am an error');
> log.error('Some context', "This error has just happened: %s", err);
ERR! Some context Error: I am an error
ERR! Some context     at repl:1:11
ERR! Some context     at sigintHandlersWrap (vm.js:22:35)
ERR! Some context     at sigintHandlersWrap (vm.js:96:12)
ERR! Some context     at ContextifyScript.Script.runInThisContext (vm.js:21:12)
ERR! Some context     at REPLServer.defaultEval (repl.js:313:29)
ERR! Some context     at bound (domain.js:280:14)
ERR! Some context     at REPLServer.runBound [as eval] (domain.js:293:12)
ERR! Some context     at REPLServer.<anonymous> (repl.js:513:10)
ERR! Some context     at emitOne (events.js:101:20)
ERR! Some context     at REPLServer.emit (events.js:188:7)
ERR! Some context  This error has just happened: %s Error: I am an error
ERR! Some context     at repl:1:11
ERR! Some context     at sigintHandlersWrap (vm.js:22:35)
ERR! Some context     at sigintHandlersWrap (vm.js:96:12)
ERR! Some context     at ContextifyScript.Script.runInThisContext (vm.js:21:12)
ERR! Some context     at REPLServer.defaultEval (repl.js:313:29)
ERR! Some context     at bound (domain.js:280:14)
ERR! Some context     at REPLServer.runBound [as eval] (domain.js:293:12)
ERR! Some context     at REPLServer.<anonymous> (repl.js:513:10)
ERR! Some context     at emitOne (events.js:101:20)
ERR! Some context     at REPLServer.emit (events.js:188:7)
```